src/main/java/net/glowstone/net/handler/play/player/PlayerAbilitiesHandler.java

```diff
@@ -14,8 +16,14 @@ public void handle(GlowSession session, PlayerAbilitiesMessage message) {
         // other values should match what we've sent in the past but are ignored here
 
         GlowPlayer player = session.getPlayer();
+        if (!player.getAllowFlight()) {
```

Comment:
Hm. This would prevent the player from disabling flight if they aren't allowed to fly in the first place.

This packet could also contain other information than just flight status, so returning here would break that.

I think the logic should look something like this:

1. If the flying flag is the same as the current flight status, don't do anything.
2. If the flying flag is different, and
   i. The flying flag is set to *true*, and the player is allowed to fly, then fire the event
   ii. The flying flag is set to *false*, fire the event
3. Check if the event is cancelled. If
   i. the event is cancelled, send the abilities to the player and return (sync)
   ii. the event is not cancelled, call the setFlying method